### PR TITLE
AVRO-1401: @Nullable does not work with byte[]

### PR DESCRIPTION
**Pull request by nandorKollar**

### CODE REVIEW

File: lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java [hunk 1]
```diff
@@ -181,13 +181,15 @@ protected boolean isRecord(Object datum) {
 
   /**
    * Returns true also for non-string-keyed maps, which are written as an array
-   * of key/value pair records.
+   * of key/value pair records. Returns false for array of bytes, since it should
```
Comment (zivanfi) [inline lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java]: I find it a bit strange that the javadoc comment only documents the exception cases and not the main behavior. I would suggest the following instead:

> Returns true for arrays and false otherwise, with the following exceptions:
> * Returns true for non-string-keyed maps, which are written as an array of key/value pair records.
> * Returns false for arrays of bytes, since those should be treated as byte data type instead.

File: lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java [hunk 2]
```diff
@@ -181,13 +181,15 @@ protected boolean isRecord(Object datum) {
...
+   * be treated as byte data type instead.
    */
   @Override
   protected boolean isArray(Object datum) {
     if (datum == null) return false;
+    Class c = datum.getClass();
     return (datum instanceof Collection)
-      || datum.getClass().isArray()
+      || (c.isArray() && !(c.getComponentType() == Byte.TYPE))
```
Comment (zivanfi) [inline lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java]: I would suggest `c.getComponentType() != Byte.TYPE` instead of `!(c.getComponentType() == Byte.TYPE)`

File: lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflect.java [hunk 3]
```diff
@@ -1048,4 +1050,17 @@ public void testAvroDefault() {
           +"{\"name\":\"foo\",\"type\":\"int\",\"default\":1}]}");
   }
 
+  private static class NullableBytesTest {
+    @Nullable
+    byte[] bytes = "foo".getBytes();
+  }
+
+  @Test
+  public void testNullableByteArray() throws IOException {
```
Comment (zivanfi) [inline lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflect.java]: Since there are no asserts in this test, could you document in a comment how this test would fail without the fix? Thanks!

Comment (nandorKollar) [inline lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflect.java]: Instead of documenting the test, I rewrote it. It asserts for the output in the updated version

**Comment by nandorKollar** [comment]
@zivanfi thanks for you comments. Updated the pull request, could you please have a look at it and share you opinion?

File: lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java [hunk 4]
```diff
@@ -180,14 +180,18 @@ protected boolean isRecord(Object datum) {
   }
 
   /**
-   * Returns true also for non-string-keyed maps, which are written as an array
-   * of key/value pair records.
+   * Returns true for arrays and false otherwise, with the following exceptions:
+   * <ul>
+   * <li><p>Returns true for non-string-keyed maps, which are written as an array of key/value pair records.</p></li>
```
Comment (zivanfi) [inline lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java]: Nit: No need for the `<p>`-s inside the `<li>`-s.

File: lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java [hunk 5]
```diff
@@ -180,14 +180,18 @@ protected boolean isRecord(Object datum) {
...
+   * <li><p>Returns false for arrays of bytes, since those should be treated as byte data type instead.</p></li>
+   * <ul>
```
Comment (zivanfi) [inline lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java]: This should be `</ul>`

**Comment by nandorKollar** [comment]
Fixed closing tag for list.

File: lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflect.java [hunk 6]
```diff
@@ -1048,4 +1048,19 @@ public void testAvroDefault() {
...
+    @Override
+    public boolean equals(Object obj) {
+      return obj instanceof NullableBytesTest
+              && Arrays.equals(((NullableBytesTest) obj).bytes, this.bytes);
+    }
+  }
+
+  @Test
+  public void testNullableByteArray() throws Exception {
+    checkReadWrite(new NullableBytesTest());
```
Comment (gszadovszky) [inline lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflect.java]: As we are testing a nullable field I would also test the scenario when this field is really null.

Comment (nandorKollar) [inline lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflect.java]: @gszadovszky updated the pul request with an additional test case to cover null field value.
